### ISSUE TITLE
Cookie protection should not work across containers

### ISSUE THREAD

**Issue by practik** [mrdokenny/Cookie-AutoDelete#262]
Closing a tab in one container does not delete that tab's cookies if the same domain is still open in a different container. That doesn't make sense to me. Since the purpose of containers is to isolate users' activities, CAD should respect that isolation. When deciding whether to delete a closed tab's cookies, CAD should only check for open tabs in the same container.

### Steps to reproduce

1. In your Firefox preferences, make sure you have at least one container set up besides the default container. If you're creating a new container, initialize your CAD settings for it as described [here](https://github.com/Cookie-AutoDelete/Cookie-AutoDelete/issues/197#issuecomment-337523756).
2. Choose a domain (`example.com`) that normally stores cookies and is not whitelisted in your CAD prefs. Enable automatic cleaning with no delay in your CAD prefs.
3. Make sure you don't have any tabs open to `example.com`.
4. Make sure your Firefox prefs are set to allow cookies from `example.com`, and if you have any `example.com` cookies stored already, delete them.
5. Open a new tab (Tab A) in any container, including the default container, and go to `example.com`.
6. Check to make sure that `example.com` has actually stored cookies, and then clear them, either with the CAD "Clear cookies for this domain" button or manually in your Firefox prefs.
7. Open a new tab (Tab B) in any other container and go to `example.com`.
8. Check to make sure that `example.com` has stored cookies again, and then close Tab B.

Expected results: CAD instantly deletes `example.com` cookies from Container B's cookie store.
Actual results: CAD deletes `example.com` cookies from Container B's store only after Tab A is closed.

### Your settings 

- OS/version: macOS 10.12.6
- Browser/version: FF 57.0
- Cookie AutoDelete version: 2.0.1
All settings checked except Show Notification After Cookie Cleanup?

**Comment by practik** [comment]
Just a quick follow-up:

When I opened this issue last week, I wasn't sure it really mattered. In the STR example above, I could also easily argue that since the `example.com` cookies in Container B are theoretically isolated from the `example.com` tab in Container A, who cares if they don't get deleted when you close Tab B?

But the idea still bugged me, and now I've figured out one reason why: If you later open a new `example.com` tab in Container B, the leftover cookies are still there, waiting to continue tracking you.

**Comment by jomo** [comment]
This concerns me as well. I have a container where I want to permanently stay logged in to certain domains, and thus have a whitelist rule *for that container*. Also, I usually have a tab pinned/open with of those domains all the time. I expect tabs in other (or no) containers to delete their cookies after I close the tab. But when I never close the container tab with the whitelist rule, the cookies for that site are never deleted in other containers.

**Comment by Blizzke** [comment]
Yeah I have the same with for example google related services. I have a google container where I'm logged in to gmail for example, but any google tab outside of this container should have its cookies deleted asap.

**Comment by echosa** [comment]
Just want to chime in here to say I also experience this issue. I have a similar experience to @Blizzke where I have a container specifically for Google services, but open tabs in that container keep Google cookies from being deleted in my other containers.

**Comment by SpaceClicker** [comment]
I was wondering if there is something blocking this "enhancement" or if it is just low on the priorities list.
If I have google.com open in two tabs, container1 and container2, and I close the tab with container2, it will not delete the cookies in container2 until I also close the tab with container1.

This means that if I always have a tab open for google,.com in one container (say the default container), cookies will never get deleted in any of the other containers for google.com.  This is not good.

Is it possible to delete the cookies when all tabs for a given "domain & container" is closed instead of just when all tabs for a given domain are closed?

Also, the same thing for domain change.

Thanks!

**Comment by kennethtran93** [comment]
Hi all,

It's a low priority for now.  Better/more logic in the back end is needed to also filter by containers.

**Comment by mike-tn** [comment]
+1 vote for cleaning cookies for a given "domain & container"!
This is hi priority feature!
(ForgetMeNot add-on can do it already!)

**Comment by kennethtran93** [comment]
Those who know how to test it please head on over to #753 to grab the dev files.
